fix emacs pylint error when running with tramp

ramiro050 opened the pull request:
## Description

The function `tramp-dissect-file-name` now returns a `tramp-file-name` structure instead of an array. Therefore, the function `aref` can no longer be used to extract the `filename` from a tramp file. This fix correctly extracts the `filename` from the `tramp-file-name` structure by using the emacs built-in function `cl-struct-slot-value`.

## Type of Changes

<!-- Leave the corresponding lines for the applicable type of change: -->

|     | Type                   |
| --- | ---------------------- |
| ✓   | :bug: Bug fix          |

Pierre-Sassoulas reviewed (changes requested):
Hi thank you for the merge request ! Could you add a test for this ? So we do not break it later on :) Also you can autoformat the code by using pre-commit:
```bash
pip install -r requirement_tests.txt
pre-commit install
```

ramiro050 commented:
I changed my fix to be much simpler. Instead of using the function `cl-struct-slot-value`, I now use the `tramp` built-in function `tramp-file-local-name` to get the filename. Now the fix is a very simple `if` statement, and it should not break with future changes to `tramp` unless they significantly change the `tramp` interface. 

Is there testing already in `pylint` for `elisp` code? I couldn't find any. I don't think my fix can be tested from `epylint` because it has to execute the code in emacs.

Pierre-Sassoulas reviewed (commented):
Yeah there is no tests fort that and up until now there wasn't any change tout this directory. I'm not confortable merging a change I donc 't understandwithout tests. How is this used donc you need elisp in the packaged pylint ? Because we just removed it in #4294

ramiro050 commented:
# What is `TRAMP`

In emacs, `tramp` is a way to seemlessly work on remote computers. For
example, you can access files over ssh like any other file by searching
for it as `/ssh:user@host:/path/to/file`. Similarly, many function that
analyze files in emacs (such as `pylint` or `compile`) recognize when
the file is remote and execute the function in the remote computer.

## What is broken?

When sending the shell command to execute to the remote computer, the
localname of the file in the remote computer must be used. This is what
is not working currently.

## Is `elisp` needed in the `pylint` package?

No. All that emacs needs is the `pylint.el` file in the `elisp` folder.
The user is expected to install `pylint` separate. All that `pylint.el`
does is it sends shell commands to `pylint`, gets the output, parses it
to make it interactive, such as adding links to the lines where warnings
and errors are happening, and does some syntax highlighting of the
output.

Every time there is a merge on the `pylint` main branch, the file
`pylint.el` is automaticaally compiled and made available to emacs users
through the package manager Melpa.
[Here](https://github.com/melpa/melpa/blob/master/recipes/pylint) is the
recipe used by Melpa.

# Why `pylint.el` change is needed

In 2012, a major redesign of `pylint.el` was made
([here](https://github.com/PyCQA/pylint/commit/3c42bad4a87ae4445732c2e0f2c23b52c4399bbf#diff-f295ba37b663ba92dcbfa8980e291517a624afa479a3334a114e6a5bbb3cd95f)).
In this change was when the function I am interested in fixing was
added. At the time the function did not have issues because the way that
a `tramp-file-name` structure was represented in memory was as an array.
However, in 2017, a major redesign of `tramp.el` was made
([here](https://git.savannah.gnu.org/cgit/tramp.git/commit/?id=c7e57ee53191054c7defb4e6c360a068833ece33)).
Most notably, the way that the `tramp-file-name` was represented changed
from a `vector` to a `list`, as shown in this `diff` from the link:

``` elisp
+;; The basic structure for remote file names.  We use a list,
+;; otherwise the test in `tramp-cache-data' fails.
+(cl-defstruct (tramp-file-name (:type list) :named)
+  method user domain host port localname hop)
```

Rough translation to `python`:

``` python
class TrampFileName:
    def __init__(self, method, user, domain, host, port, localname, hop):
        self.filename = [method, user, domain, host, port, localname, hop]
```

In the same `tramp` commit linked above, there is also this deletion:

``` elisp
-(defun tramp-file-name-localname (vec)
-  "Return localname component of VEC."
-  (and (tramp-file-name-p vec) (aref vec 3)))
```

If you look at the `pylint.el` part I am trying to modify, you see that
this is the same code written for getting the `localname`. However, it
was removed from `tramp.el` because the structure is no longer a vector,
so it doesn't work.

# My changes

I have made a different set of changes to `pylint.el` that try to keep
the structure as similar as possible. Instead of `aref`, I am now using
`elt`, which is the equivalent of `aref` for lists ([documentation
here](https://www.gnu.org/software/emacs/manual/html_node/elisp/Sequence-Functions.html)).
The only other thing I added is the following line

``` elisp
  (let* ((filename (buffer-file-name))
+        (localname-offset (cl-struct-slot-offset 'tramp-file-name 'localname))
         (filename (or (and (tramp-tramp-file-p filename)
```

The way that `let` statements work is that they take a pair `(variable
value)` and assign the `variable` to the value. In `python` this would
be

``` python
filename = buffer_file_name()
localname_offset = TrampFileName.offsetof("localname")
filename = ...
```

Here is the code being run line by line in the REPL. Notice that `aref` fails because the `tramp-flie-name` is no longer a `vector`.

<img width="689" alt="Screen Shot 2021-04-11 at 9 41 03 AM" src="https://user-images.githubusercontent.com/20114526/114314192-8cabfa00-9aae-11eb-96bc-b93847b73d90.png">

I have tested this change on my machine and it works as expected, as can be seen below:

<img width="691" alt="Screen Shot 2021-04-11 at 10 00 26 AM" src="https://user-images.githubusercontent.com/20114526/114314226-b5cc8a80-9aae-11eb-818a-577c31b1963c.png">

Pierre-Sassoulas reviewed (approved):
Thank you for the detailed explanation :) Let me know when the MR is ready I'll merge it.

ramiro050 commented:
I'm glad the explanation was helpful! I just double checked everything, and it should be good to merge

Pierre-Sassoulas commented:
Thanks you for your contribution and for explaining in so much details :)
